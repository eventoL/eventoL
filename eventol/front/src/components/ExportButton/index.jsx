import _ from 'lodash';
import React from 'react';
import PropTypes from 'prop-types';
import {CSVLink} from 'react-csv';

import './index.scss';

export default class ExportButton extends React.PureComponent {
  static propTypes = {
    data: PropTypes.arrayOf(PropTypes.shape()),
    filename: PropTypes.string,
    label: PropTypes.string,
    type: PropTypes.oneOf([
      'default',
      'primary',
      'success',
      'info',
      'warning',
      'danger',
      'link',
    ]),
  };

  static defaultProps = {
    data: [],
    filename: 'export',
    label: '',
    type: 'default',
  };

  state = {
    header: [],
    rows: [],
    totals: [],
  };

  getField = (columns, field) => {
    const values = [];
    columns.forEach(column => {
      if (!column.hasOwnProperty('Header')) return null;
      if (!column.hasOwnProperty('columns')) values.push(column[field]);
      else column.columns.forEach(subcolumn => values.push(subcolumn[field]));
      return column;
    });
    return values;
  };

  getRows(columns) {
    const {data} = this.props;
    const accessors = this.getField(columns, 'accessor');
    return data.map(row =>
      accessors.map(accessor => this.runAccessor(accessor, row))
    );
  }

  runAccessor = (accessor, row) => {
    if (typeof accessor === 'string') return row[accessor];
    return accessor(row);
  };

  updateCsv(columns) {
    const {data} = this.props;
    if (!_.isEmpty(data)) {
      const header = this.getField(columns, 'Header');
      const rows = this.getRows(columns);
      const totals = this.getField(columns, 'total');
      this.setState({header, rows, totals});
    }
  }

  render() {
    const {type, label, filename} = this.props;
    const {header, rows, totals} = this.state;
    const csvData = [header, ...rows, totals];
    return (
      <div className={`export btn btn-raised btn-${type}`}>
        <CSVLink data={csvData} filename={`${filename}.csv`}>
          {label}
        </CSVLink>
      </div>
    );
  }
}
