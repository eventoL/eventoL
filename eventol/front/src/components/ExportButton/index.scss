@import '../../styles/variables';

.export {
  a {
    color: $a-color;
    text-decoration: none;
  }
}
