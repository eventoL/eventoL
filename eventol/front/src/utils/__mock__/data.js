const place1 = {
  address_components: [
    {long_name: '1667', short_name: '1667', types: ['street_number']},
    {long_name: 'Ayacucho', short_name: 'Ayacucho', types: ['route']},
    {
      long_name: 'Centro',
      short_name: 'Centro',
      types: ['neighborhood', 'political'],
    },
    {
      long_name: 'Rosario',
      short_name: 'Rosario',
      types: ['locality', 'political'],
    },
    {
      long_name: 'Departamento Rosario',
      short_name: 'Departamento Rosario',
      types: ['administrative_area_level_2', 'political'],
    },
    {
      long_name: 'Santa Fe',
      short_name: 'Santa Fe',
      types: ['administrative_area_level_1', 'political'],
    },
    {long_name: 'Argentina', short_name: 'AR', types: ['country', 'political']},
    {long_name: 'S2000', short_name: 'S2000', types: ['postal_code']},
    {long_name: 'BYS', short_name: 'BYS', types: ['postal_code_suffix']},
  ],
  formatted_address: 'Ayacucho 1667, S2000BYS Rosario, Santa Fe, Argentina',
  geometry: {
    location: {lat: -32.9586923, lng: -60.629729},
    location_type: 'ROOFTOP',
    viewport: {
      south: -32.9600412802915,
      west: -60.63107798029148,
      north: -32.9573433197085,
      east: -60.628380019708516,
    },
  },
  place_id: 'ChIJ7QvrMwCrt5URpyBml6doLnY',
  plus_code: {
    compound_code: '29RC+G4 Rosario, Santa Fe, Argentina',
    global_code: '47VX29RC+G4',
  },
  types: ['street_address'],
};

const place2 = {
  address_components: [
    {long_name: '1551', short_name: '1551', types: ['street_number']},
    {long_name: 'Sarmiento', short_name: 'Sarmiento', types: ['route']},
    {
      long_name: 'San Nicolás',
      short_name: 'San Nicolás',
      types: ['political', 'sublocality', 'sublocality_level_1'],
    },
    {
      long_name: 'Comuna 1',
      short_name: 'Comuna 1',
      types: ['administrative_area_level_2', 'political'],
    },
    {
      long_name: 'Buenos Aires',
      short_name: 'CABA',
      types: ['administrative_area_level_1', 'political'],
    },
    {long_name: 'Argentina', short_name: 'AR', types: ['country', 'political']},
    {long_name: 'C1042', short_name: 'C1042', types: ['postal_code']},
    {long_name: 'ABC', short_name: 'ABC', types: ['postal_code_suffix']},
  ],
  formatted_address: 'Sarmiento 1551, C1042ABC CABA, Argentina',
  geometry: {
    location: {lat: -34.6052436, lng: -58.388569700000005},
    location_type: 'ROOFTOP',
    viewport: {
      south: -34.6065925802915,
      west: -58.38991868029149,
      north: -34.60389461970851,
      east: -58.387220719708466,
    },
  },
  place_id: 'ChIJbXpzCcTKvJURbeIVAyh6SnU',
  plus_code: {
    compound_code: '9JV6+WH Buenos Aires, Argentina',
    global_code: '48Q39JV6+WH',
  },
  types: ['street_address'],
};

export const event1 = {
  attendees: 10,
  backdrop: 'event1backdrop',
  eventSlug: 'event1slug',
  overview: 'event1overview',
  place: JSON.stringify(place1),
  tags: [
    {
      name: 'slug-name',
      slug: 'slug-slug',
    },
  ],
  title: 'event 1',
  url: 'event1Url',
  id: 1,
};

export const event2 = {
  attendees: 25,
  backdrop: 'event2backdrop',
  eventSlug: 'event2slug',
  overview: 'event2overview',
  place: JSON.stringify(place2),
  tags: [
    {
      name: 'slug-name',
      slug: 'slug-slug',
    },
  ],
  title: 'event 2',
  url: 'event2Url',
  id: 2,
};

export const events = [event1, event2];

export default {
  event1,
  event2,
  events,
};
