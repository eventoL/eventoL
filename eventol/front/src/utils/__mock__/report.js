const getConfirmationReport = (confirmed, notConfirmed) => ({
  confirmed,
  not_confirmed: notConfirmed,
  total: confirmed + notConfirmed,
});

export const usersReports = {
  attendee: {
    with_event_user: getConfirmationReport(1, 2),
    without_event_user: getConfirmationReport(2, 3),
  },
  collaborator: getConfirmationReport(3, 4),
  event_user: getConfirmationReport(4, 5),
  installer: getConfirmationReport(6, 7),
  organizer: getConfirmationReport(7, 8),
  speakers: 9,
};

export const activitiesReport = {
  level_count: {
    '1': 3,
    '2': 2,
    '3': 1,
  },
  status_count: {
    '1': 2,
    '2': 1,
  },
  type_count: {
    Talk: 4,
    Workshop: 2,
  },
  ...getConfirmationReport(1, 2),
};

export const installationsReport = {
  hardware_count: {
    notebook: 1,
    phone: 2,
  },
  software_count: {
    ubuntu: 3,
    fedora: 2,
    arch: 1,
  },
  installer_count: {
    '1': 2,
    '2': 1,
  },
  total: 2,
};

export const eventReport = {
  // users data
  ...usersReports,

  // activities data
  activity: activitiesReport,

  // installations data
  installation: installationsReport,
};

export const event1PrivateData = {
  id: 1,
  email: 'organizer1@mailinator.com',
  organizers: 'FirstName1 LastName1 Username1',
};

export const event2PrivateData = {
  id: 2,
  email: 'organizer2@mailinator.com',
  organizers: 'FirstName2 LastName2 Username2',
};

export const eventsPrivateData = [
  {...event1PrivateData},
  {...event2PrivateData},
];

export const getEventData = (id, name, tags, report, eventSlug, location) => ({
  id,
  name,
  tags,
  report,
  location,
  event_slug: eventSlug,
});

export const event1Data = getEventData(
  1,
  'event name',
  {name: 'tag name', slug: 'tag slug'},
  eventReport,
  'event_slug',
  ['Capital Federal', 'Pompeya', 'Buenos Aires', 'Argentina']
);

export const event2Data = getEventData(
  2,
  'event 2 name',
  {name: 'tag 2 name', slug: 'tag 2 slug'},
  eventReport,
  'event_2_slug',
  ['Capital Federal', 'Monserrat', 'Buenos Aires', 'Argentina']
);

export const eventsData = [{...event1Data}, {...event2Data}];

export const expectedActivityDetail = {
  levels: {'1': 6, '2': 4, '3': 2},
  status: {'1': 4, '2': 2},
  types: {Talk: 8, Workshop: 4},
};

export const expectedInstallationsSoftwares = {arch: 2, fedora: 4, ubuntu: 6};

export const expectedTotals = {
  activities: {
    confirmed: 2,
    details: expectedActivityDetail,
    not_confirmed: 4,
    total: 6,
  },
  attendees: {confirmed: 6, not_confirmed: 10, total: 16},
  collaborators: {confirmed: 6, not_confirmed: 8, total: 14},
  installations: {softwares: expectedInstallationsSoftwares, total: 4},
  installers: {confirmed: 12, not_confirmed: 14, total: 26},
  organizers: {confirmed: 14, not_confirmed: 16, total: 30},
  speakers: 18,
};

export const expectedEventDataWithoutPrivateData = {
  ...event1Data,
  assistanceDetail: {attendees: {confirmed: 3, not_confirmed: 5, total: 8}},
  locationDetail: {
    address: 'Pompeya',
    address_detail: 'Capital Federal',
    province: 'Buenos Aires',
  },
};

export const expectedEventDataWithPrivateData = {
  ...expectedEventDataWithoutPrivateData,
  ...event1PrivateData,
};

export const expectedValues = {
  expectedTotals,
  expectedActivityDetail,
  expectedInstallationsSoftwares,
  expectedEventDataWithPrivateData,
  expectedEventDataWithoutPrivateData,
};

const mockedData = {
  activitiesReport,
  eventData: event1Data,
  event1Data,
  event2Data,
  eventsData,
  eventReport,
  event1PrivateData,
  event2PrivateData,
  eventsPrivateData,
  getEventData,
  installationsReport,
  usersReports,
  expectedValues,
};
export default mockedData;
