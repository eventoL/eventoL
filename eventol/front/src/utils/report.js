import _ from 'lodash';

export const parseInstallationSoftwares = allData => {
  const softwares = new Set();
  allData.forEach(event =>
    Object.keys(event.report.installation.software_count).forEach(key =>
      softwares.add(key)
    )
  );
  const installationsSoftwares = {};
  softwares.forEach(software => {
    const sum = _.sumBy(
      allData,
      `report.installation.software_count.${software}`
    );
    installationsSoftwares[software] = sum;
  });
  return installationsSoftwares;
};

export const parseActivitiesDetails = allData => {
  const status = new Set();
  const types = new Set();
  const levels = new Set();
  allData.forEach(event => {
    Object.keys(event.report.activity.status_count).forEach(key =>
      status.add(key)
    );
    Object.keys(event.report.activity.type_count).forEach(key =>
      types.add(key)
    );
    Object.keys(event.report.activity.level_count).forEach(key =>
      levels.add(key)
    );
  });
  const activityDetail = {types: {}, status: {}, levels: {}};
  status.forEach(element => {
    const sum = _.sumBy(allData, `report.activity.status_count.${element}`);
    activityDetail.status[element] = sum;
  });
  types.forEach(type => {
    const sum = _.sumBy(allData, `report.activity.type_count.${type}`);
    activityDetail.types[type] = sum;
  });
  levels.forEach(level => {
    const sum = _.sumBy(allData, `report.activity.level_count.${level}`);
    activityDetail.levels[level] = sum;
  });
  return activityDetail;
};

export const parseTotals = allData => {
  const totals = {
    speakers: _.sumBy(allData, 'report.speakers'),
    attendees: {
      confirmed:
        _.sumBy(allData, 'report.attendee.with_event_user.confirmed') +
        _.sumBy(allData, 'report.attendee.without_event_user.confirmed'),
      not_confirmed:
        _.sumBy(allData, 'report.attendee.with_event_user.not_confirmed') +
        _.sumBy(allData, 'report.attendee.without_event_user.not_confirmed'),
      total:
        _.sumBy(allData, 'report.attendee.with_event_user.total') +
        _.sumBy(allData, 'report.attendee.without_event_user.total'),
    },
    organizers: {
      confirmed: _.sumBy(allData, 'report.organizer.confirmed'),
      not_confirmed: _.sumBy(allData, 'report.organizer.not_confirmed'),
      total: _.sumBy(allData, 'report.organizer.total'),
    },
    collaborators: {
      confirmed: _.sumBy(allData, 'report.collaborator.confirmed'),
      not_confirmed: _.sumBy(allData, 'report.collaborator.not_confirmed'),
      total: _.sumBy(allData, 'report.collaborator.total'),
    },
    installers: {
      confirmed: _.sumBy(allData, 'report.installer.confirmed'),
      not_confirmed: _.sumBy(allData, 'report.installer.not_confirmed'),
      total: _.sumBy(allData, 'report.installer.total'),
    },
    activities: {
      confirmed: _.sumBy(allData, 'report.activity.confirmed'),
      not_confirmed: _.sumBy(allData, 'report.activity.not_confirmed'),
      total: _.sumBy(allData, 'report.activity.total'),
      details: parseActivitiesDetails(allData),
    },
    installations: {
      total: _.sumBy(allData, 'report.installation.total'),
      softwares: parseInstallationSoftwares(allData),
    },
  };
  return totals;
};

export const parseEvent = (event, eventsPrivateData = []) => {
  const {
    location,
    report: {attendee},
  } = event;
  const eventData = {
    locationDetail: {
      address_detail: location.slice(0, -3).join(' '),
      address: location[location.length - 3],
      province: location[location.length - 2],
    },
    assistanceDetail: {
      attendees: {
        confirmed:
          attendee.with_event_user.confirmed +
          attendee.without_event_user.confirmed,
        not_confirmed:
          attendee.with_event_user.not_confirmed +
          attendee.without_event_user.not_confirmed,
        total:
          attendee.with_event_user.total + attendee.without_event_user.total,
      },
    },
    ...event,
  };
  if (!_.isEmpty(eventsPrivateData)) {
    const privateData = _.find(eventsPrivateData, {id: event.id});
    if (!_.isEmpty(privateData)) return {...privateData, ...eventData};
  }
  return eventData;
};
