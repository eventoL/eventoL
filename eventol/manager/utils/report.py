import logging

logger = logging.getLogger("eventol")


def count_by(elements, getter, increment=None):
    return_dict = {}
    for element in elements:
        try:
            field = str(getter(element))
            if field in return_dict:
                return_dict[field] += increment(element) if increment else 1
            else:
                return_dict[field] = increment(element) if increment else 1
        except Exception as error:
            logger.error(error)
    return return_dict
