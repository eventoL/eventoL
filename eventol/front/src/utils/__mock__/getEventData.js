import {eventReport} from './report';

export const getEventData = (id, name, tags, report, eventSlug, location) => ({
  id,
  name,
  tags,
  report,
  location,
  event_slug: eventSlug,
});
export const event1Data = getEventData(
  1,
  'event name',
  {name: 'tag name', slug: 'tag slug'},
  eventReport,
  'event_slug',
  ['Capital Federal', 'Pompeya', 'Buenos Aires', 'Argentina']
);
export const event2Data = getEventData(
  2,
  'event 2 name',
  {name: 'tag 2 name', slug: 'tag 2 slug'},
  eventReport,
  'event_2_slug',
  ['Capital Federal', 'Monserrat', 'Buenos Aires', 'Argentina']
);
export const eventsData = [{...event1Data}, {...event2Data}];
