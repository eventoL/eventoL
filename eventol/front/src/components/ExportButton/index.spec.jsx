import wait from 'waait';
import React from 'react';
import renderer from 'react-test-renderer';

jest.mock('react-csv', () => ({CSVLink: 'CSVLink'}));

import ExportButton from '.';
import {eventsData} from '../../utils/__mock__/report';

describe('ExportButton', () => {
  let tree;
  let element;
  let columns;
  let instance;
  let component;

  beforeEach(() => {
    columns = [
      {
        Header: 'Event',
        columns: [
          {
            Footer: 'Footer',
            Header: 'Name',
            accessor: 'name',
            minWidth: 150,
            resizable: false,
            sortable: true,
            total: 'Events: 1',
          },
        ],
      },
      {columns: []},
      {
        Footer: 'Footer id',
        Header: 'ID',
        accessor: 'id',
        minWidth: 150,
        resizable: false,
        sortable: true,
        total: 'Ids: 1',
      },
      {
        Header: 'Tag',
        columns: [
          {
            Footer: 'Footer 2',
            Header: 'Tag',
            accessor: ({tags: {name}}) => name,
            minWidth: 200,
            resizable: false,
            sortable: true,
            total: 'Tags: 1',
          },
          {
            Footer: 'Footer 3',
            Header: 'Slug',
            accessor: ({tags: {slug}}) => slug,
            minWidth: 100,
            resizable: false,
            sortable: true,
            total: 'Slugs: 1',
          },
        ],
      },
    ];
  });

  describe('Without data', () => {
    beforeEach(() => {
      element = (
        <ExportButton
          data={[]}
          filename="export"
          label="Label"
          type="default"
        />
      );
      component = renderer.create(element);
      tree = component.toJSON();
      instance = component.getInstance();
    });

    test('Default', async () => {
      expect(tree).toMatchSnapshot();

      expect(instance.state.header).toEqual([]);
      expect(instance.state.rows).toEqual([]);
      expect(instance.state.totals).toEqual([]);

      instance.updateCsv(columns);

      component.update(element);
      await wait(0);
      tree = component.toJSON();
      expect(tree).toMatchSnapshot();

      expect(instance.state.header).toEqual([]);
      expect(instance.state.rows).toEqual([]);
      expect(instance.state.totals).toEqual([]);
    });
  });

  describe('With data', () => {
    beforeEach(() => {
      element = (
        <ExportButton
          data={eventsData}
          filename="export"
          label="Label"
          type="default"
        />
      );
      component = renderer.create(element);
      tree = component.toJSON();
      instance = component.getInstance();
    });

    test('Default', async () => {
      expect(tree).toMatchSnapshot();

      expect(instance.state.header).toEqual([]);
      expect(instance.state.rows).toEqual([]);
      expect(instance.state.totals).toEqual([]);

      instance.updateCsv(columns);

      component.update(element);
      await wait(0);
      tree = component.toJSON();
      expect(tree).toMatchSnapshot();

      expect(instance.state.header).toEqual(['Name', 'ID', 'Tag', 'Slug']);
      expect(instance.state.rows).toEqual([
        ['event name', 1, 'tag name', 'tag slug'],
        ['event 2 name', 2, 'tag 2 name', 'tag 2 slug'],
      ]);
      expect(instance.state.totals).toEqual([
        'Events: 1',
        'Ids: 1',
        'Tags: 1',
        'Slugs: 1',
      ]);
    });
  });
});
