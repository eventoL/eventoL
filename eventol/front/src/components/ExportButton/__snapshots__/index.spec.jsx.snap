// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ExportButton With data Default 1`] = `
<div
  className="export btn btn-raised btn-default"
>
  <CSVLink
    data={
      Array [
        Array [],
        Array [],
      ]
    }
    filename="export.csv"
  >
    Label
  </CSVLink>
</div>
`;

exports[`ExportButton With data Default 2`] = `
<div
  className="export btn btn-raised btn-default"
>
  <CSVLink
    data={
      Array [
        Array [
          "Name",
          "ID",
          "Tag",
          "Slug",
        ],
        Array [
          "event name",
          1,
          "tag name",
          "tag slug",
        ],
        Array [
          "event 2 name",
          2,
          "tag 2 name",
          "tag 2 slug",
        ],
        Array [
          "Events: 1",
          "Ids: 1",
          "Tags: 1",
          "Slugs: 1",
        ],
      ]
    }
    filename="export.csv"
  >
    Label
  </CSVLink>
</div>
`;

exports[`ExportButton Without data Default 1`] = `
<div
  className="export btn btn-raised btn-default"
>
  <CSVLink
    data={
      Array [
        Array [],
        Array [],
      ]
    }
    filename="export.csv"
  >
    Label
  </CSVLink>
</div>
`;

exports[`ExportButton Without data Default 2`] = `
<div
  className="export btn btn-raised btn-default"
>
  <CSVLink
    data={
      Array [
        Array [],
        Array [],
      ]
    }
    filename="export.csv"
  >
    Label
  </CSVLink>
</div>
`;
